### PR TITLE
Test branch for fixing 1 sample shift

### PR DESCRIPTION
**Pull request by derick-hess**
Has anything changed on the fdsn webservices side?  This part of the code hasn't been touched since this fix was done last july and verified with a bunch of different web service requests that matched.

removing the bug fix seems to fix the shift. Though it returns 1 sample less than it should at then end of the request, but that can also be fixed.

@nick-iris I'm going to run a bunch more tests, can you do the same so we can verify?

### CODE REVIEW

**Comment by nick-iris** [comment]
This fix only works because of the following addtional correction in ph5toms. I don't think this correction should be ph5toms, but instead in ph5toapi. 

https://github.com/PIC-IRIS/PH5/blob/master/ph5/clients/ph5toms.py#L406-L411
```
# if start time is before requested start time move up 1 sample
# and delete first sample of data
if trace.start_time.epoch() < stc.starttime:
    trace.start_time = trace.start_time + \
                                        1 / float(stc.sample_rate))
    trace.data = trace.data[1:]
```

**Comment by derick-hess** [comment]
That should be in ph5api for sure. I think the best way to fix that is to remove the original bug fix and move this code into ph5 api

**Comment by derick-hess** [comment]
Actually im digging deeper to find where this could be. I'm now thinking the real issue isn't in cut but in ph5_g_receivers.read_trace I'll keep going. I want to fix this so we don't have to do anythign specifically f in the case of nodes

**Comment by derick-hess** [comment]
It's a rounding error here:  https://github.com/PIC-IRIS/PH5/blob/master/ph5/core/ph5api.py#L920
This fixes it while letting us remove the unnecessary code from ph5toms and specific node references in ph5api.

I"m going to look for potential other rounding errors in cut.

Basically what was happening is it was rounding the requested sample to start at sample 41 in the window instead of sample 42.  The floating point number was 4.99999

**Comment by nick-iris** [comment]
Thanks for tracking this down.

Chad mentioned to me that there is potential for a loss of precision when using floating point values for epoch times (especially for high frequency data). For this reason, in the current version of libmseed, epoch times are represented as integers. For a long term fix, we'll want to ensure that no times are represented as floats.

**Comment by derick-hess** [comment]
Yeah  so instead of epoch in seconds we do it epoch in milliseconds.  That makes sense and is something we should move to

**Comment by chad-iris** [comment]
I suggest doing epoch time in nanoseconds as a integer.  libmseed currently does epoch in microseconds but the next generation miniSEED is going to nanosecond resolution.  It is particularly justified when handling high rate (small sample period) data.

**Comment by derick-hess** [comment]
I'll add the epoch time change in nanoseconds to the list of changes that we will be giving to the PH5 committee.

**Comment by derick-hess** [comment]
I left the following in to ph5toms so that output should match exactly what the fdsn web services output.  Otherwise it will return 1 less sample at the end compared to the fdsn web service.
 
` new_endtime = stc.endtime + (1 / float(stc.sample_rate))`

**Comment by derick-hess** [comment]
I'm working on creating a very small PH5 data set we can either download from the passcal cloud server or include in github for testing. I'm thinking just 10 seconds of data from a node, an rt130, and an rt125 should do. The rt130 will include various sample rates as well.

**Comment by nick-iris** [comment]
I just tested an LHZ (YW 503 01 LHZ) channel and unfortunately the two services now produce different results in the milliseconds for this channel. Before the recent code change they matched identically. 

Please compare the output from 

irisws-timeseries:
http://service.iris.edu/irisws/timeseries/1/query?network=YW&station=503&location=01&channel=LHZ&starttime=2016-06-23T16:43:00&endtime=2016-06-23T16:45:00&output=miniseed

with

`ph5toms -p /hdf5-data2/PH5_Experiments/pn4/16-015/ -n master.ph5 --format=mseed --station=503 --channel=LHZ --starttime=2016-06-23T16:43:58.0 --stoptime=2016-06-23T16:43:59.0`

**Comment by derick-hess** [comment]
I think I found it.  I'll do some tests on all teh different instrument types and sample rates to make sure they all match before i post the fix

**Comment by derick-hess** [comment]
Now i'm looking at the data.  I will need to get the raw data at work tomorrow but I am finding differences in the data I think. 

For instance the fdsn webservices station 1001 at 1466553600 (GMT: Wednesday, June 22, 2016 12:00:00 AM) has a sample entered twice the sample value is 423 at 1466553600 and the same sample repeated the next sample. The PH5 miniseed output doesn't have this sample duplicated.

I'll check tomorrow if this is the same in the PH5 data before it is touched by any software and in the raw node data.

**Comment by derick-hess** [comment]
I can confirm when cutting around 466553600 (GMT: Wednesday, June 22, 2016 12:00:00 AM) in station 1001 DP1 in fdsn web service that data is duplicated in the fdsn aNd for all time i checked prior to that the data matched exactly and after that it differs by  a sample for the times i checked after that,

In the morning I will check the raw data to confirm

**Comment by nick-iris** [comment]
The LHZ channels now match, but the DP* channels no longer do. Please compare the following data:

https://service.iris.edu/irisws/timeseries/1/query?network=YW&station=1002&location=--&channel=DPZ&starttime=2016-06-23T16:43:00.0&endtime=2016-06-23T16:45:00.0&output=miniseed

`ph5toms -n master.ph5 -p /hdf5-data2/PH5_Experiments/pn4/16-015/ --station=1002 --channel=DPZ --starttime=2016-06-23T16:43:00.0 --stoptime=2016-06-23T16:45:00.0`

**Comment by derick-hess** [comment]
I know the DP channels don't match. There is an extra sample in the DMC miniseed at midnight of the first recording day (at least on 1001 and I suspect on all of them). I've checked the  data inside PH5 and they match and PH5 is picking the correct sample and times. I verified this by also hand calculating what sample should be picked and looking it up in the PH5 data arrays.

It very unlikely that samples back to back have the same value and even more suspect that it happens exactly at midnight of the first day of recording. I'm still trying to get the raw node data to confirm directly from the rg16 file but I'm fairly certain that PH5 is correct in this case.

File: ph5/core/ph5api.py [hunk 1]
```diff
@@ -910,37 +910,38 @@ def cut (self, das, start_fepoch, stop_fepoch, chan=1, sample_rate=None, apply_t
...
+
             #   Requested start before start of window, we must need to start cutting at start of window
             if start_fepoch < window_start_fepoch :
                 cut_start_fepoch = window_start_fepoch
                 cut_start_sample = 0
             else :
             #   Cut start is somewhere in window
                 cut_start_fepoch = start_fepoch
-                cut_start_sample = int ((start_fepoch - window_start_fepoch) * sr)
+                cut_start_sample = round((((start_fepoch*1000000 - window_start_fepoch*1000000))/1000000 * sr))
```
Comment (nick-iris) [inline ph5/core/ph5api.py]: Why do you multiply by 1000000 here?

Comment (derick-hess) [reply ph5/core/ph5api.py]: That doesn't really need to be there.  That was probably left there when I was testing where rounding errors were occurring. We should probably change that back before creating the release.

Comment (nick-iris) [reply ph5/core/ph5api.py]: Okay thanks. Can you please change it back?

Comment (derick-hess) [reply ph5/core/ph5api.py]: Sure I have a meeting right now, but should have it fixed within the hour

Comment (nick-iris) [reply ph5/core/ph5api.py]: Thanks Derick.

**Pull request merged**